src/harness/testcases/grant_testcase.py

```diff
@@ -624,3 +624,245 @@ def test_WINNF_FT_S_GRA_28(self):
+    the CBSD which is inside the protection zone of Federal Incumbent.
+    """
+
+    # Register the device inside the protection zone
+    device_inside_esc_zone = json.load(
+        open(os.path.join('testcases', 'testdata', 'device_c.json')))
+    self._sas_admin.InjectFccId({'fccId': device_inside_esc_zone['fccId']})
+    request = {'registrationRequest': [device_inside_esc_zone]}
+    response = self._sas.Registration(request)['registrationResponse'][0]
+    # Check registration response
+    self.assertEqual(response['response']['responseCode'], 0)
+    cbsd_id = response['cbsdId']
+    del request, response
+    # Create and trigger the ESC Zone
+    exclusion_zone = json.load(
+        open(os.path.join('testcases', 'testdata', 'exclusion_zone_0.json')))
+    zone_response = self._sas_admin.InjectEscZone(exclusion_zone)
+    trigger_esc_zone = {'zoneId': zone_response['zoneId'],
+                                    'frequencyRange': {
+                                     'lowFrequency': 3620000000.0,
+                                     'highFrequency': 3630000000.0}}
+    trigger_id = self._sas_admin.TriggerEscZone(trigger_esc_zone)
+    # Request grant
+    grant_0 = json.load(
+        open(os.path.join('testcases', 'testdata', 'grant_0.json')))
+    grant_0['cbsdId'] = cbsd_id
+    request = {'grantRequest': [grant_0]}
+    response = self._sas.Grant(request)['grantResponse'][0]
+    # Check grant response
+    self.assertEqual(response['cbsdId'], cbsd_id)
+    if(response['response']['responseCode'] == 400) :
+      self.assertFalse('grantId' in response)
+    else:
+      self.assertTrue('grantId' in response)
+      self.assertEqual(response['channelType'], 'GAA')
+      grant_id = response['grantId']
+      del request, response
+      request = {
+      'heartbeatRequest': [{'cbsdId': cbsd_id,'grantId': grant_id,'operationState': 'GRANTED'}]}
+      response = self._sas.Heartbeat(request)['heartbeatResponse'][0]
+      # Check the heartbeat response
+      self.assertEqual(response['cbsdId'], cbsd_id)
+      self.assertEqual(response['grantId'], grant_id)
+      self.assertFalse('transmitExpireTime' in response)
+      self.assertEqual(response['response']['responseCode'], 501)
+
+  @winnforum_testcase 
+  def test_WINFF_FT_S_GRA_4(self):
+    """Successful CBSD grant request.
+        No incumbent present in the PAL frequency range requested by the CBSD.
+    """
+
+    # Register device inside PPA
+    device_a = json.load(
+        open(os.path.join('testcases', 'testdata', 'device_a.json')))
+    self._sas_admin.InjectFccId({'fccId': device_a['fccId']})
+    request = {'registrationRequest': [device_a]}
+    response = self._sas.Registration(request)['registrationResponse'][0]
+    # Check registration response
+    self.assertEqual(response['response']['responseCode'], 0)
+    cbsd_id = response['cbsdId']
+    del request, response
+    # Inject PAL Database Recoord
```

Comment:
I think we also need to inject Census Tracts Zone Object and based on the zoneId the palId would need the FIPS Code from the zoneId as well as licenseAreaExtent will have Census Tract Zone Id in it.